internal/app/machined/pkg/system/runner/process/process.go

```diff
@@ -78,27 +90,95 @@ func (p *processRunner) Close() error {
 }
 
 type commandWrapper struct {
-	cmd              *exec.Cmd
+	launcher         *cap.Launcher
+	ctty             optional.Optional[int]
+	cgroupFile       *os.File
+	stdin            *os.File
+	stdout           *os.File
+	stderr           *os.File
 	afterStart       func()
 	afterTermination func() error
 }
 
+func dropCaps(droppedCapabilities []string, launcher *cap.Launcher) error {
+	droppedCaps := strings.Join(droppedCapabilities, ",")
+
+	if droppedCaps != "" {
+		caps := strings.Split(droppedCaps, ",")
+		dropCaps := xslices.Map(caps, func(c string) cap.Value {
+			capability, capErr := cap.FromName(c)
+			if capErr != nil {
+				panic(fmt.Errorf("failed to parse capability: %s", capErr))
+			}
+
+			return capability
+		})
+
+		iab := cap.IABGetProc()
+		if err := iab.SetVector(cap.Bound, true, dropCaps...); err != nil {
+			return fmt.Errorf("failed to set capabilities: %w", err)
+		}
+
+		launcher.SetIAB(iab)
+	}
+
+	return nil
+}
+
+// This callback is run in the thread before executing child process.
+func beforeExecCallback(pa *syscall.ProcAttr, data any) error {
+	wrapper, ok := data.(*commandWrapper)
+	if !ok {
+		return fmt.Errorf("failed to get command info")
+	}
+
+	ctty, cttySet := wrapper.ctty.Get()
+	if cttySet {
+		if pa.Sys == nil {
+			pa.Sys = &syscall.SysProcAttr{}
+		}
+
+		pa.Sys.Ctty = ctty
+		pa.Sys.Setsid = true
+		pa.Sys.Setctty = true
+	}
+
+	pa.Files = []uintptr{
+		wrapper.stdin.Fd(),
```

Comment:
In case the caller doesn't specify the stdin file to use, we end up with nil -> fd 0 here, so the child will be given Talos' own stdin. Should we override that with /dev/null?

Reply:
i think we need nil checks then, otherwise calling methods on `os.File` would panic :thinking:

Reply:
We only add calls to .Close in case we opened the file successfully, and .Fd doesn't fail but returns 0 (stdin) in case of being called on nil